### PR TITLE
vmcluster: re-routing enhancement

### PR DESCRIPTION
**Pull request by zekker6**
This PR implements re-routing enhancement approach suggested here - https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4922#issuecomment-1768146878

Test results are shared here - https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4922#issuecomment-1790640274

### CODE REVIEW

File: lib/ingestserver/conns_map.go [hunk 1]
```diff
@@ -37,10 +39,28 @@ func (cm *ConnsMap) Delete(c net.Conn) {
...
 	cm.mu.Lock()
+	var connCloseInterval time.Duration
+	conns := len(cm.m)
+	if conns == 0 {
+		connCloseInterval = 0
+	} else {
+		connCloseInterval = time.Duration(grace.Milliseconds()/int64(conns)) * time.Millisecond
+	}
+
+	// Sort addresses in order to make the order of closing connections deterministic.
```
Comment (hagen1778) [inline lib/ingestserver/conns_map.go:52]:
```suggestion
	// Sort addresses in order to make the order of closing connections deterministic across vmstorages.
```

File: app/vmstorage/servers/vminsert.go [hunk 2]
```diff
@@ -134,7 +136,7 @@ var (
 )
 
 // MustStop gracefully stops s so it no longer touches s.storage after returning.
-func (s *VMInsertServer) MustStop() {
+func (s *VMInsertServer) MustStop(grace time.Duration) {
```
Comment (hagen1778) [inline app/vmstorage/servers/vminsert.go:139]: nit: would be nice to log that there is a grace shutdown and that vmstorage is going to close connections one-by-one on the time interval

File: lib/ingestserver/conns_map.go [hunk 3]
```diff
@@ -37,10 +39,28 @@ func (cm *ConnsMap) Delete(c net.Conn) {
 }
 
 // CloseAll closes all the added conns.
-func (cm *ConnsMap) CloseAll() {
+func (cm *ConnsMap) CloseAll(grace time.Duration) {
 	cm.mu.Lock()
+	var connCloseInterval time.Duration
```
Comment (hagen1778) [inline lib/ingestserver/conns_map.go:44]: should we have a fast path if: grace==0 or len(cm.m) < 2 ?

File: docs/CHANGELOG.md [hunk 4]
```diff
@@ -77,6 +77,7 @@ The sandbox cluster installation is running under the constant load generated by
 * FEATURE: dashboards/vmalert: add new panel `Missed evaluations` for indicating alerting groups that miss their evaluations.
 * FEATURE: all: track requests with wrong auth key and wrong basic auth at `vm_http_request_errors_total` [metric](https://docs.victoriametrics.com/#monitoring) with `reason="wrong_auth_key"` and `reason="wrong_basic_auth"`. See [this issue](https://github.com/victoriaMetrics/victoriaMetrics/issues/4590). Thanks to @venkatbvc for the [pull request](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/5166).
 * FEATURE: [vmauth](https://docs.victoriametrics.com/vmauth.html): add `tls_insecure_skip_verify` parameter which allows to disable TLS verification for backend connection. See [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/5240).
+* FEATURE: `vmstorage`: add `-storage.gracefulShutdownDuration` command-line flag which can be used for enhancing re-routing performance during `vmstorage` restarts. See [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4922) and [these docs](https://docs.victoriametrics.com/Cluster-VictoriaMetrics.html#improving-re-routing-performance-during-restart) for details.
```
Comment (hagen1778) [inline docs/CHANGELOG.md:80]:
```suggestion
* FEATURE: `vmstorage`: add `-storage.gracefulShutdownDuration` command-line flag for gradual connections termination between vminserts and vmstorage. This should reduce the impact generated by re-routing data samples from terminated vmstorage node to the remaining nodes. Please note, the `-storage.gracefulShutdownDuration` has a default value to enable graceful shutdown of vmstorage nodes. To disable this behavior, set `-storage.gracefulShutdownDuration=0s`. See [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4922) and [these docs](https://docs.victoriametrics.com/Cluster-VictoriaMetrics.html#improving-re-routing-performance-during-restart) for details.
```

File: docs/Cluster-VictoriaMetrics.md [hunk 5]
```diff
@@ -470,8 +470,28 @@ This strategy allows upgrading the cluster without downtime if the following con
 - The updated config / upgraded binary is compatible with the remaining components in the cluster.
   See the [CHANGELOG](https://docs.victoriametrics.com/CHANGELOG.html) for compatibility notes between different releases.
 
-  If at least a single condition isn't met, then the rolling restart may result in cluster unavailability
-  during the config update / version upgrade. In this case the following strategy is recommended.
+If at least a single condition isn't met, then the rolling restart may result in cluster unavailability
+during the config update / version upgrade. In this case the following strategy is recommended.
+
+#### Improving re-routing performance during restart
```
Comment (hagen1778) [inline docs/Cluster-VictoriaMetrics.md:476]: nit: I think we don't need extra section here - it should be listed as a bullet-point  in `Updating / reconfiguring cluster nodes`

Comment (hagen1778) [reply docs/Cluster-VictoriaMetrics.md:476]: I'd also recommend to make it prompt: just state what vmstorage does and how user can action on this. Plus link to the issue.

**Review by valyala — APPROVED** [review]

**Comment by valyala** [comment]
@zekker6 , thanks for the great improvement to rolling restart in VictoriaMetrics cluster!

**Pull request merged**